diff --git a/vllm/model_executor/layers/quantization/__init__.py b/vllm/model_executor/layers/quantization/__init__.py
index 1cb23e7a1..fa20bb0f7 100644
--- a/vllm/model_executor/layers/quantization/__init__.py
+++ b/vllm/model_executor/layers/quantization/__init__.py
@@ -35,6 +35,11 @@ QuantizationMethods = Literal[
     "moe_wna16",
     "torchao",
     "auto-round",
+    # NOTE: MLU quantization methods
+    "smoothquant",
+    "weightonly",
+    "awq_mlu",
+    "gptq_mlu",
 ]
 QUANTIZATION_METHODS: list[str] = list(get_args(QuantizationMethods))
 

