diff --git a/vllm/v1/engine/__init__.py b/vllm/v1/engine/__init__.py
index 59463f1ba..2b2a5d057 100644
--- a/vllm/v1/engine/__init__.py
+++ b/vllm/v1/engine/__init__.py
@@ -55,7 +55,7 @@ class EngineCoreRequest(
     arrival_time: float
     lora_request: Optional[LoRARequest]
     cache_salt: Optional[str]
-    data_parallel_rank: Optional[int]
+    data_parallel_rank: Optional[int] = None
 
     # Index of the client, used to ensure outputs are sent back to the same
     # client for this request when scaling out the front-end.

