diff --git a/vllm_mlu/vllm_mlu/distributed/kv_transfer/kv_connector/factory.py b/vllm_mlu/vllm_mlu/distributed/kv_transfer/kv_connector/factory.py
new file mode 100644
index 000000000..d70c3eeae
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/distributed/kv_transfer/kv_connector/factory.py
@@ -0,0 +1,17 @@
+from vllm.distributed.kv_transfer.kv_connector.factory import KVConnectorFactory
+
+
+MLUKVConnectors: dict[str, tuple[str, str]] = {
+    "MLUSharedStorageConnector": (
+        "vllm_mlu.distributed.kv_transfer.kv_connector.v1.shared_storage_connector",
+        "SharedStorageConnector"
+    ),
+    "MLUNixlConnector": (
+        "vllm_mlu.distributed.kv_transfer.kv_connector.v1.nixl_connector",
+        "MLUNixlConnector"
+    ),
+}
+
+for name, (module_path, class_name) in MLUKVConnectors.items():
+    if name not in KVConnectorFactory._registry:
+        KVConnectorFactory.register_connector(name, module_path, class_name)
\ No newline at end of file

