diff --git a/vllm/forward_context.py b/vllm/forward_context.py
index f3b0518a4..e07b30ea8 100644
--- a/vllm/forward_context.py
+++ b/vllm/forward_context.py
@@ -122,8 +122,9 @@ def set_forward_context(attn_metadata: Any,
     if need_to_track_batchsize:
         forward_start_time = time.perf_counter()
     dp_metadata: Optional[DPMetadata] = None
-    if vllm_config.parallel_config.data_parallel_size > 1 and (
-            attn_metadata is not None or num_tokens is not None):
+    if (vllm_config.parallel_config.data_parallel_size > 1 and (
+            attn_metadata is not None or num_tokens is not None)
+            and not vllm_config.mlu_config.enable_custom_data_parallel_opt):
         dp_metadata = DPMetadata.make(vllm_config.parallel_config,
                                       attn_metadata, num_tokens or 0,
                                       num_tokens_across_dp)

