diff --git a/vllm_mlu/tools/build.property b/vllm_mlu/tools/build.property
new file mode 100644
index 000000000..2412d020e
--- /dev/null
+++ b/vllm_mlu/tools/build.property
@@ -0,0 +1,10 @@
+VLLM_VERSION=0.9.1+mlu0.11.0.pt27
+TORCH_MLU_OPS_VERSION=1.8.0+torch2.7.1
+CATCH_VERSION=1.29.1+torch2.7.1
+CNCL_VERSION=1.29.2-1
+CNNL_VERSION=2.1.9-1
+CNNLEXTRA_VERSION=2.2.7-1
+CNTOOLKIT_VERSION=4.4.2-1
+CNCLEP_VERSION=1.0.0-1
+MLUOPS_VERSION=1.8.0-1
+TRITON_VERSION=3.2.0+mlu1.7.0

