diff --git a/vllm_mlu/vllm_mlu/v1/core/single_type_kv_cache_manager.py b/vllm_mlu/vllm_mlu/v1/core/single_type_kv_cache_manager.py
new file mode 100644
index 000000000..1cb1d82da
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/v1/core/single_type_kv_cache_manager.py
@@ -0,0 +1,11 @@
+from vllm.v1.core.single_type_kv_cache_manager import (
+    FullAttentionManager, SlidingWindowManager, spec_manager_map)
+
+from vllm_mlu.v1.kv_cache_interface import (MLUFullAttentionSpec,
+                                            MLUSlidingWindowSpec)
+
+spec_manager_map.update({
+    MLUFullAttentionSpec: FullAttentionManager,
+    MLUSlidingWindowSpec: SlidingWindowManager,
+})
+

