diff --git a/vllm_mlu/vllm_mlu/model_executor/models/__init__.py b/vllm_mlu/vllm_mlu/model_executor/models/__init__.py
new file mode 100755
index 000000000..3a40748c0
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/model_executor/models/__init__.py
@@ -0,0 +1,131 @@
+from vllm import ModelRegistry
+
+
+def register_model():
+    from .baichuan import MLUBaichuanForCausalLM  # noqa: F401
+    from .baichuan import MLUBaiChuanForCausalLM  # noqa: F401
+    from .bloom import MLUBloomForCausalLM  # noqa: F401
+    from .chatglm import MLUChatGLMForCausalLM  # noqa: F401
+    from .falcon import MLUFalconForCausalLM  # noqa: F401
+    from .gpt_neox import MLUGPTNeoXForCausalLM  # noqa: F401
+    from .hunyuan import MLUHunYuanForCausalLM  # noqa: F401
+    from .internvl import MLUInternVLChatModel  # noqa: F401
+    from .internlm2 import MLUInternLM2ForCausalLM  # noqa: F401
+    from .llama import MLULlamaForCausalLM  # noqa: F401
+    from .llama_eagle import MLUEagleLlamaForCausalLM  # noqa: F401
+    from .mixtral import MLUMixtralForCausalLM  # noqa: F401
+    from .phi3 import MLUPhi3ForCausalLM  # noqa: F401
+    from .qwen import MLUQWenLMHeadModel  # noqa: F401
+    from .qwen2_5_vl import MLUQwen2_5_VLForConditionalGeneration  # noqa: F401
+    from .qwen2_moe import MLUQwen2MoeForCausalLM  # noqa: F401
+    from .qwen2_vl import MLUQwen2VLForConditionalGeneration  # noqa: F401
+    from .qwen2 import MLUQwen2ForCausalLM  # noqa: F401
+    from .qwen3_moe import MLUQwen3MoeForCausalLM  # noqa: F401
+    from .qwen3 import MLUQwen3ForCausalLM  # noqa: F401
+    from .deepseek_v2 import MLUDeepseekV2ForCausalLM  # noqa: F401
+    from .step3_vl import MLUStep3VLForConditionalGeneration
+    from .step3_text import MLUStep3TextForCausalLM
+    from .glm4_moe import MLUGlm4MoeForCausalLM  # noqa: F401
+
+    ModelRegistry.register_model(
+        "BaichuanForCausalLM",
+        "vllm_mlu.model_executor.models.baichuan:MLUBaichuanForCausalLM")
+    ModelRegistry.register_model(
+        "BaiChuanForCausalLM",
+        "vllm_mlu.model_executor.models.baichuan:MLUBaiChuanForCausalLM")
+    ModelRegistry.register_model(
+        "BloomForCausalLM",
+        "vllm_mlu.model_executor.models.bloom:MLUBloomForCausalLM")
+    ModelRegistry.register_model(
+        "ChatGLMModel",
+        "vllm_mlu.model_executor.models.chatglm:MLUChatGLMForCausalLM")
+    ModelRegistry.register_model(
+        "FalconForCausalLM",
+        "vllm_mlu.model_executor.models.falcon:MLUFalconForCausalLM")
+    ModelRegistry.register_model(
+        "GPTNeoXForCausalLM",
+        "vllm_mlu.model_executor.models.gpt_neox:MLUGPTNeoXForCausalLM")
+    ModelRegistry.register_model(
+        "HunYuanForCausalLM",
+        "vllm_mlu.model_executor.models.hunyuan:MLUHunYuanForCausalLM")
+    ModelRegistry.register_model(
+        "InternLM2ForCausalLM",
+        "vllm_mlu.model_executor.models.internlm2:MLUInternLM2ForCausalLM")
+    ModelRegistry.register_model(
+        "InternVLChatModel",
+        "vllm_mlu.model_executor.models.internvl:MLUInternVLChatModel")
+    ModelRegistry.register_model(
+        "DeepseekV3ForCausalLM",
+        "vllm_mlu.model_executor.models.deepseek_v2:MLUDeepseekV2ForCausalLM")
+    ModelRegistry.register_model(
+        "DeepseekV2ForCausalLM",
+        "vllm_mlu.model_executor.models.deepseek_v2:MLUDeepseekV2ForCausalLM")
+    ModelRegistry.register_model(
+        "LlamaForCausalLM",
+        "vllm_mlu.model_executor.models.llama:MLULlamaForCausalLM")
+    ModelRegistry.register_model(
+        "InternLMForCausalLM",
+        "vllm_mlu.model_executor.models.llama:MLULlamaForCausalLM")
+    ModelRegistry.register_model(
+        "LLaMAForCausalLM",
+        "vllm_mlu.model_executor.models.llama:MLULlamaForCausalLM")
+    ModelRegistry.register_model(
+        "MistralModel",
+        "vllm_mlu.model_executor.models.llama:MLULlamaForCausalLM")
+    ModelRegistry.register_model(
+        "EagleLlamaForCausalLM",
+        "vllm_mlu.model_executor.models.llama_eagle:MLUEagleLlamaForCausalLM")
+    ModelRegistry.register_model(
+        "MistralForCausalLM",
+        "vllm_mlu.model_executor.models.llama:MLULlamaForCausalLM")
+    ModelRegistry.register_model(
+        "MixtralForCausalLM",
+        "vllm_mlu.model_executor.models.mixtral:MLUMixtralForCausalLM")
+    ModelRegistry.register_model(
+        "QuantMixtralForCausalLM",
+        "vllm_mlu.model_executor.models.mixtral:MLUMixtralForCausalLM")
+    ModelRegistry.register_model(
+        "Phi3ForCausalLM",
+        "vllm_mlu.model_executor.models.phi3:MLUPhi3ForCausalLM")
+    ModelRegistry.register_model(
+        "QWenLMHeadModel",
+        "vllm_mlu.model_executor.models.qwen:MLUQWenLMHeadModel")
+    ModelRegistry.register_model(
+        "Qwen2_5_VLForConditionalGeneration",
+        "vllm_mlu.model_executor.models.qwen2_5_vl:MLUQwen2_5_VLForConditionalGeneration")
+    ModelRegistry.register_model(
+        "Qwen2MoeForCausalLM",
+        "vllm_mlu.model_executor.models.qwen2_moe:MLUQwen2MoeForCausalLM")
+    ModelRegistry.register_model(
+        "Qwen2VLForConditionalGeneration",
+        "vllm_mlu.model_executor.models.qwen2_vl:MLUQwen2VLForConditionalGeneration")
+    ModelRegistry.register_model(
+        "Qwen2ForCausalLM",
+        "vllm_mlu.model_executor.models.qwen2:MLUQwen2ForCausalLM")
+    ModelRegistry.register_model(
+        "Qwen2ForSequenceClassification",
+        "vllm_mlu.model_executor.models.qwen2:MLUQwen2ForCausalLM")
+    ModelRegistry.register_model(
+        "Qwen3ForCausalLM",
+        "vllm_mlu.model_executor.models.qwen3:MLUQwen3ForCausalLM")
+    ModelRegistry.register_model(
+        "Qwen3MoeForCausalLM",
+        "vllm_mlu.model_executor.models.qwen3_moe:MLUQwen3MoeForCausalLM")
+    ModelRegistry.register_model(
+        "DeepseekV3ForCausalLM",
+        "vllm_mlu.model_executor.models.deepseek_v2:MLUDeepseekV2ForCausalLM")
+    ModelRegistry.register_model(
+        "DeepseekV2ForCausalLM",
+        "vllm_mlu.model_executor.models.deepseek_v2:MLUDeepseekV2ForCausalLM")
+    ModelRegistry.register_model(
+        "Step3TextForCausalLM",
+        "vllm_mlu.model_executor.models.step3_text:MLUStep3TextForCausalLM")
+    ModelRegistry.register_model(
+        "Step3VLForConditionalGeneration",
+        "vllm_mlu.model_executor.models.step3_vl:MLUStep3VLForConditionalGeneration")
+    ModelRegistry.register_model(
+        "Glm4MoeForCausalLM",
+        "vllm_mlu.model_executor.models.glm4_moe:MLUGlm4MoeForCausalLM")
+    ModelRegistry.register_model(
+        "Glm4vMoeForConditionalGeneration",
+        "vllm_mlu.model_executor.models.glm4_1v:MLUGlm4vMoeForConditionalGeneration")

