diff --git a/vllm_mlu/vllm_mlu/model_executor/models/phi3.py b/vllm_mlu/vllm_mlu/model_executor/models/phi3.py
new file mode 100644
index 000000000..3f249542d
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/model_executor/models/phi3.py
@@ -0,0 +1,13 @@
+from vllm_mlu.model_executor.models.llama import MLULlamaForCausalLM
+
+
+class MLUPhi3ForCausalLM(MLULlamaForCausalLM):
+
+    packed_modules_mapping = {
+        "qkv_proj": [
+            "qkv_proj",
+        ],
+        "gate_up_proj": [
+            "gate_up_proj",
+        ],
+    }
\ No newline at end of file

