diff --git a/vllm/engine/protocol.py b/vllm/engine/protocol.py
index 727d59283..ee73b9c5f 100644
--- a/vllm/engine/protocol.py
+++ b/vllm/engine/protocol.py
@@ -57,6 +57,7 @@ class EngineClient(ABC):
         trace_headers: Optional[Mapping[str, str]] = None,
         prompt_adapter_request: Optional[PromptAdapterRequest] = None,
         priority: int = 0,
+        data_parallel_rank: Optional[int] = None,
     ) -> AsyncGenerator[RequestOutput, None]:
         """Generate outputs for a request."""
         ...

