diff --git a/vllm_mlu/vllm_mlu/v1/worker/block_table.py b/vllm_mlu/vllm_mlu/v1/worker/block_table.py
new file mode 100644
index 000000000..77be6bd18
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/v1/worker/block_table.py
@@ -0,0 +1,69 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+import torch
+
+from vllm.v1.worker.block_table import BlockTable
+from vllm.logger import init_logger
+
+from vllm_mlu.mlu_hijack_utils import MluHijackObject
+
+logger = init_logger(__name__)
+
+
+class BlockTable_MluHijack(BlockTable):
+
+    def __init__(
+        self,
+        max_num_reqs: int,
+        max_num_blocks_per_req: int,
+        max_num_batched_tokens: int,
+        pin_memory: bool,
+        device: torch.device,
+    ):
+        self.max_num_reqs = max_num_reqs
+        self.max_num_blocks_per_req = max_num_blocks_per_req
+        self.max_num_batched_tokens = max_num_batched_tokens
+        self.pin_memory = pin_memory
+        self.device = device
+
+        self.block_table = torch.zeros(
+            (max_num_reqs, max_num_blocks_per_req),
+            device=self.device,
+            dtype=torch.int32,
+        )
+        self.block_table_cpu = torch.zeros(
+            (max_num_reqs, max_num_blocks_per_req),
+            device="cpu",
+            dtype=torch.int32,
+            pin_memory=pin_memory,
+        )
+        self.block_table_np = self.block_table_cpu.numpy()
+        self.num_blocks_per_row = np.zeros(max_num_reqs, dtype=np.int32)
+
+        '''
+        =============================
+        Modify by vllm_mlu
+        =============================
+        @brief: change slot_mapping dtype for int64 to int32
+        '''
+        self.slot_mapping_cpu = torch.zeros(self.max_num_batched_tokens,
+                                            dtype=torch.int32,
+                                            device="cpu",
+                                            pin_memory=self.pin_memory)
+        self.slot_mapping_np = self.slot_mapping_cpu.numpy()
+        self.slot_mapping = torch.zeros(self.max_num_batched_tokens,
+                                        dtype=torch.int32,
+                                        device=self.device)
+        '''
+        ==================
+        End of MLU Hijack
+        ==================
+        '''
+
+
+MluHijackObject.apply_hijack(
+    BlockTable,
+    BlockTable.__init__,
+    BlockTable_MluHijack.__init__
+)
\ No newline at end of file

